diff --git a/src/native/api.c b/src/native/api.c
index 36f42fd..41b9436 100644
--- a/src/native/api.c
+++ b/src/native/api.c
@@ -138,7 +138,7 @@ llhttp_errno_t llhttp_finish(llhttp_t* parser) {
 
   switch (parser->finish) {
     case HTTP_FINISH_SAFE_WITH_CB:
-      CALLBACK_MAYBE(parser, on_message_complete);
+      SPAN_CALLBACK_MAYBE(parser, on_message_complete, "", 0);
       if (err != HPE_OK) return err;
 
     /* FALLTHROUGH */
@@ -322,7 +322,7 @@ int llhttp__on_headers_complete(llhttp_t* s, const char* p, const char* endp) {
 
 int llhttp__on_message_complete(llhttp_t* s, const char* p, const char* endp) {
   int err;
-  CALLBACK_MAYBE(s, on_message_complete);
+  SPAN_CALLBACK_MAYBE(s, on_message_complete, p, endp - p);
   return err;
 }
 
diff --git a/src/native/api.h b/src/native/api.h
index d86c022..dd094c3 100644
--- a/src/native/api.h
+++ b/src/native/api.h
@@ -42,7 +42,7 @@ struct llhttp_settings_s {
   llhttp_data_cb on_body;
 
   /* Possible return values 0, -1, `HPE_PAUSED` */
-  llhttp_cb      on_message_complete;
+  llhttp_data_cb on_message_complete;
 
   /* When on_chunk_header is called, the current chunk length is stored
    * in parser->content_length.
